examples/reaction-rate-learning/reaction-rate-learning.py

```diff
@@ -0,0 +1,317 @@
+# 
+# With the unknown reaction rate:
+
+# .. math::
+#    \frac{d}{dt}[C] = R([A], [B])
+
+# We denote the concentrations of A, B and C with :math:`u_1`, :math:`u_2` and
+# :math:`u_3` and add source terms :math:`f_1`, :math:`f_2` and
+# :math:`f_3` for each species, respectively. 
+# Then the coupled system of nonlinear PDEs reads:
+
+# .. math::
+#   \begin{aligned}
+#   \frac{\partial u_1}{\partial t} +
+#     w \cdot \nabla u_1 - \nabla\cdot(\epsilon\nabla u_1)
+#     &= f_1 - R(u_1, u_2) \\ 
+#     \frac{\partial u_2}{\partial t} +
+#     w \cdot \nabla u_2 - \nabla\cdot(\epsilon\nabla u_2)
+#     &= f_2 - R(u_1, u_2) \\ 
+#     \frac{\partial u_3}{\partial t} +
+#     w \cdot \nabla u_3 - \nabla\cdot(\epsilon\nabla u_3)
+#     &= f_3 + R(u_1, u_2)
+#     \end{aligned} 
+
+# Now, we model the unkown reaction rate :math:`R` with a neural network (NN)
+# :math:`R_{\theta}(u_1, u_2)` parameterized with the weights :math:`\theta` and
+# formulate the resulting hybrid PDE-NN problem as a PDE-constrained
+# optimization problem as introduced in :cite:`mitusch2021hybrid`.
+
+# Problem Definition
+# ******************
+# Assuming that we can observe the concentrations :math:`u_i^{\star}` of all
+# species i at all points in time in the whole domain, we set the goal to
+# minimize the weighted :math:`L^2` error:
+
+# .. math::
+#    \begin{aligned}
+#    \min_{\theta} \sum_{i=1,2,3} \lambda_i \int_0^T \int_{\Omega} (u_i^{\star}(x,t) - u_i^{}(x,t))^2 dxdt \\
+#    \end{aligned}
+
+
+# subject to:
+
+# .. math::
+#    \begin{aligned}
+#    \frac{\partial u_1}{\partial t} +
+#    w \cdot \nabla u_1 - \nabla\cdot(\epsilon\nabla u_1)
+#    &= f_1 - R_{\theta}(u_1, u_2)  &\text{ in } \Omega \times (0,T) \\ 
+#    \frac{\partial u_2}{\partial t} +
+#    w \cdot \nabla u_2 - \nabla\cdot(\epsilon\nabla u_2)
+#    &= f_2 - R_{\theta}(u_1, u_2)  &\text{ in } \Omega \times (0,T) \\ 
+#    \frac{\partial u_3}{\partial t} + w \cdot \nabla u_3 - \nabla\cdot(\epsilon\nabla u_3)
+#    &= f_3 + R_{\theta}(u_1, u_2) &\text{ in } \Omega \times (0,T) \\
+#    \frac{\partial u_i}{\partial n} &= 0 &\text{ on } \partial\Omega \times (0,T) \\ 
+#    \end{aligned} 
+
+# with initial conditions :math:`u_i=0` for :math:`i=1,2,3`. The weights :math:`\lambda_i`
+# of the minimization functional are introduced to account for the smaller magnitude
+# of the concentration of the newly formed species C.
+
+# FEniCS implementation
+# *********************
+
+# For the details of the implementation of the background velocity field and
+# the advection-diffusion-reaction system, we refer to
+# the `FEniCS tutorial <https://fenicsproject.org/pub/tutorial/html/._ftut1010.html#ftut1:reactionsystem>`_.
+# Here, we implement a function `solve_reaction_system` 
+# that solves the reaction system for a given reaction funtion (`reaction_func`)
+# and computes the loss for each time step, if a loss function (`loss_func`) is provided:
+
+from fenics import *
+set_log_level(50)
```

Comment:
Perhaps use the symbolic LogLevel here instead of the integer translation, i.e: `set_log_level(LogLevel.CRITICAL)` in this instance.